working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,559 @@
+  > - Where the request dictionary is missing the title
+  > - Where the request dictionary is missing the description
+</details>
+
+When we asked for help with testing these scenarios, Copilot gave us test suggestions, but they looked very different from our code in `test_routes.py`:
+```
+import pytest
+from flask import json
+
+def test_update_book_with_extra_keys(client):
+    response = client.put('/books/1', data=json.dumps({
+        "title": "New Title",
+        "description": "New Description",
+        "extra_key": "Extra Value"
+    }), content_type='application/json')
+
+    assert response.status_code == 200
+    assert b"Book #1 successfully updated" in response.data
+
+def test_update_book_without_title(client):
+    response = client.put('/books/1', data=json.dumps({
+        "description": "New Description"
+    }), content_type='application/json')
+
+    assert response.status_code == 400
+    assert b"Title is required" in response.data
+
+def test_update_book_without_description(client):
+    response = client.put('/books/1', data=json.dumps({
+        "title": "New Title"
+    }), content_type='application/json')
+
+    assert response.status_code == 400
+    assert b"Description is required" in response.data
+```
+
+We could manually update the tests to match better, but Copilot can also reformat them to be closer to what we're looking for if we follow up with a prompt like:
+
+> Please update the tests above to match the formatting used in test_routes.py
+
+The resulting test code Copilot suggests still has issues we will address, but it is structured much closer to our other tests:
+
+```
+def test_update_book_with_extra_keys(client, two_saved_books):
+    # Arrange
+    book_id = 1
+    data = {
+        "title": "Updated Book",
+        "description": "New description",
+        "extra_key": "Extra Value"
+    }
+
+    # Act
+    response = client.put(f"/books/{book_id}", json=data)
+    response_body = response.get_json()
+
+    # Assert
+    assert response.status_code == 200
+    assert response_body == "Book #1 successfully updated"
+
+def test_update_book_without_title(client, two_saved_books):
+    # Arrange
+    book_id = 1
+    data = {
+        "description": "New description"
+    }
+
+    # Act
+    response = client.put(f"/books/{book_id}", json=data)
+
+    # Assert
+    assert response.status_code == 400
+    assert b"Title is required" in response.data
+
+def test_update_book_without_description(client, two_saved_books):
+    # Arrange
+    book_id = 1
+    data = {
+        "title": "Updated Book"
+    }
+
+    # Act
+    response = client.put(f"/books/{book_id}", json=data)
+
+    # Assert
+    assert response.status_code == 400
+    assert b"Description is required" in response.data
+```
+
+If we paste these tests into our test file and run them, the test `test_update_book_with_extra_keys` passes without changes, but we need to take a deeper look at the tests for missing required data. 
```

Comment:
- you're probably picking up on my general aversion to "if" phrasing by now. To me, it sounds much more engaging to just do the thing rather than propsing a hypothetical
- explicitly add that the other tests are failing
```suggestion
Let's paste these tests into our test file and run them. The test `test_update_book_with_extra_keys` passes without changes, but the other two are failing. We need to take a deeper look at the tests for missing required data.
```